index.rst

```diff
@@ -1,42 +1,198 @@
-        ...
-      module N( f, type (++) ) where
-        data family a ++ b = L a | R b
+This separation allows us to define data constructors and type
+constructors whose names coincide:
 
-The extension :extension:`ExplicitNamespaces` is implied by
-:extension:`TypeOperators` and (for some reason) by :extension:`TypeFamilies`.
+.. code:: haskell
 
-In addition, with :extension:`PatternSynonyms` you can prefix the name of a
-data constructor in an import or export list with the keyword
-``pattern``, to allow the import or export of a data constructor without
-its parent type constructor.
+   data T = T
+
+(Similarly, with the ``TypeOperators`` extension it is possible to define
+term-level operators and type families whose names coincide.)
+
+At use sites, GHC infers which ``T`` is referred to from the context:
+
+.. code:: haskell
+
+   t :: T  -- type-level T
+   t = T   -- term-level T
+
+The use of identical names for type-level and term-level entities is called
+*punning*.  Haskell makes heavy use of punning in its built-in syntax and common
+types.
+
+However, there are various contexts in which an occurrence of a name may refer
+either to the type namespace or the data namespace, and it is not always clear
+which is meant. In particular, the following may mention both type-level and
+term-level entities:
+
+- Import and export lists
+
+- Fixity declarations
+
+- ``WARNING``, ``DEPRECATED`` and ``ANN`` pragmas
+
+- Types, when using the ``DataKinds`` extension to reference a data constructor
+  at the type level
+
+- Template Haskell name quotes
+
+In certain situations, which appear to be geining prevalence in modern Haskell, precison is needed
+in specifying exactly which names are exported or imported from a module. This extension aims to
+provide precise control in the presence of puns where the same name is present in the type and
+data namespace.
+
+ExplicitNamespaces Overview
+---------------------------
+
+When the ``ExplicitNamespaces`` extensiopn is enabled, ``type`` and ``data`` keywords may be prefixed
+to a qualified constructor names in import and export lists to specify whether a type constructor
+or data constructor is being exported or brought into scope.  It is an error to use such a namespace
+specifier if the identifier is not in scope in the given namespace.
+
+More precisely, the standard grammar of Haskell import/export items is
+essentially the following (after some minor simplifications): ::
+
+      export -> qcname_ext ['(' qcname_ext_w_wildcard_1, ..., qcname_ext_w_wildcard_n ')']
+             |  'module' modid
+
+      qcname_ext_w_wildcard -> qcname_ext
+                            |  '..'
+
+      qcname_ext -> qvar
+                 |  qtycon
+
+This extension extends ``qcname_ext`` as follows: ::
+
+      qcname_ext -> qvar
+                 |  qtycon
+                 | 'type' oqtycon  -- with ExplicitNamespaces
+                 | 'data' qvarcon  -- with ExplicitNamespaces
```

Comment:
I'd suggest dropping the grammar, and simply explaining the rules more concisely.

Reply:
I would like to get some more input on this. The Haskell language is in effect being driven forward by these extensions and the user-guide description is the definitive record of the changes we are making. I don't think recent practice is a good guide here. Apart from PLD issues, as a consumer of the User's Guide, I want to see the details.